### ISSUE TITLE
Why validate on LFW with phase_train_placeholder:True?

### ISSUE THREAD

**Issue by apollo-time** [davidsandberg/facenet#74]
This is relative to #56, #72.
I check [TensorFlow-Slim image classification library Evaluation Code](https://github.com/tensorflow/models/tree/a9133ae914b44602c5f26afbbd7dd794ff9c6637/slim#Eval) and found setting is_training is to False in the [eval_image_classifier.py](https://github.com/tensorflow/models/blob/a9133ae914b44602c5f26afbbd7dd794ff9c6637/slim/eval_image_classifier.py).
`
####################
    # Select the model #
    ####################
    network_fn = nets_factory.get_network_fn(
        FLAGS.model_name,
        num_classes=(dataset.num_classes - FLAGS.labels_offset),
        is_training=False)

    ##############################################################
    # Create a dataset provider that loads data from the dataset #
    ##############################################################
    provider = slim.dataset_data_provider.DatasetDataProvider(
        dataset,
        shuffle=False,
        common_queue_capacity=2 * FLAGS.batch_size,
        common_queue_min=FLAGS.batch_size)
    [image, label] = provider.get(['image', 'label'])
    label -= FLAGS.labels_offset

    #####################################
    # Select the preprocessing function #
    #####################################
    preprocessing_name = FLAGS.preprocessing_name or FLAGS.model_name
    image_preprocessing_fn = preprocessing_factory.get_preprocessing(
        preprocessing_name,
        is_training=False)
`

So I think setting is_training with False is true when embedding for new face.
If validate on LFW with is_training is False, accuracy is low 67%.
Is it mean this model is not good trainable?

**Comment by LiuzcEECS** [comment]
I met the same problem as you did. In my opinion, when we apply batch_norm in testing, we need to use the moving_variance and moving_mean calculated from the database. However, I just printed the moving_variance and moving_mean with phase_train_placeholder: False and found the values of  moving_variance are all 1.0 .

**Comment by LiuzcEECS** [comment]
I think the decay in moving_average **was** too high (0.9997) (seems to be fixed yesterday) to record the variance and as a result, the moving_variance kept unchanged during the training. 
When we validate the model on LFW with phase_train_placeholder:True, tensorflow doesn't use the moving_variance we saved. Instead, it just calculates mean and variance of the mini-batch by `tf.nn.moments`. So my solution is to just set the phase_train_placeholder:True or you can write another batch_norm module which use `tf.nn.moments` when phase_train_placeholder is set to False instead of using the moving_variance we saved as tf.Slim does.

**Comment by apollo-time** [comment]
Thanks your reply.
Which conv layer you said moving_variance of?
And how to save moving_variance and used it in test mode?
I think extracting feature of one face with setting phase_train_placeholder:True is not good idea.
How can I extracting feature with is_train is not?

**Comment by LiuzcEECS** [comment]
1. moving_variance of **ALL** the layers in 20161030-023650 are [1.0, 1.0 , ..., 1.0] (the issue has been fixed in 20161116-234200)
2. (https://www.tensorflow.org/versions/r0.11/api_docs/python/train.html#moving-averages) provides APIs to save and restore
3. Yes, for just one face, not a good idea.
In model 20161030-023650, I think my solution will work. In model 20161116-234200, the problem has been solved. In fact, in 20161116-234200, the variable `phase_train` was set and we can't change the variable without defining a new graph. It seems that `phase_train` was set with False but I'm not sure since we can't see the the variable `phase_train` in the graph. Check the way he defines his model in `inception_resnet_v1.py` for more details.
4.We set is_train with False to use the moving_variables we saved. And you are right, in my last solution, we don't need to care about whether is_train is True or False. By the way, I **was** going to apply the solution and I saw the update of the model released on 20161116 :)

**Comment by apollo-time** [comment]
Thanks very much. I'll test the new model 20161116  soon.!!!

**Comment by apollo-time** [comment]
And how did you trained the new mode?
Is this Classifier-training-of-inception-resnet-v1 method Ok with the last source?

**Comment by LiuzcEECS** [comment]
Sorry, I haven't tried the new training method yet. I think I'll try to figure it out this week.

**Comment by apollo-time** [comment]
If you have  tried the new method, please share the result for me.
Thanks!!!

**Comment by davidsandberg** [comment]
Closing due to inactivity. Open if needed.
